File: java/56_MergeIntervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        if (intervals.length < 2) {
            return intervals;
        }

        List<int[]> res = new ArrayList<>();
        Arrays.sort(intervals, (arr1, arr2) -> Integer.compare(arr1[0], arr2[0]));

        int[] cur = intervals[0];
        res.add(cur);

        for (int[] interval : intervals) {
            int curEnd = cur[1];
            int nextStart = interval[0];
            int nextEnd = interval[1];
            if (curEnd >= nextStart) {
                cur[1] = Math.max(nextEnd, curEnd);
            } else {
                res.add(interval);
                cur = interval;
            }
        }

        return res.toArray(new int[res.size()][]);
    }
}

File: c++/56_MergeIntervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& a) {
        vector<vector<int>> res;
        if (a.empty()) return res;
        sort(a.begin(), a.end());
        int l = a[0][0], r = a[0][1];
        for (int i = 1; i < a.size(); i++) {
            if (r < a[i][0]) {
                res.push_back({l, r});
                l = a[i][0], r = a[i][1];
            } else {
                r = max(r, a[i][1]);
            }
        }
        res.push_back({l, r});
        return res;
    }
};